File: tests/runtime/resource-borrow/runner.go
package runner

import (
	"context"
	"fmt"

	wrpc "wrpc.io/go"

	"driver/runner/test/resource_borrow/to_test"
)

func Run(ctx context.Context, c wrpc.Invoker) error {
	thing := must(to_test.NewThing(ctx, c, 42))
	assertEqual(must(to_test.Foo(ctx, c, thing.Borrow())), 42+1+2)
	return nil
}

func must[T any](v T, err error) T {
	if err != nil {
		panic(err)
	}
	return v
}

func assertEqual[T comparable](a T, b T) {
	if a != b {
		panic(fmt.Sprintf("%v not equal to %v", a, b))
	}
}

File: tests/runtime/resource-borrow/runner.rs
use crate::client::test::resource_borrow::to_test::{foo, Thing};

pub async fn run(
    wrpc: &impl ::wit_bindgen_wrpc::wrpc_transport::Invoke<Context = ()>,
) -> ::wit_bindgen_wrpc::anyhow::Result<()> {
    let thing = Thing::new(wrpc, (), 42).await?;
    assert_eq!(foo(wrpc, (), &thing.as_borrow()).await?, 42 + 1 + 2);
    Ok(())
}
